### PR TITLE
hw08

### PR DESCRIPTION
**Pull request by u2107680007911543774e7r**
check please : ((

### CODE REVIEW

File: src/main/java/homework08/BookCrossingLibrary.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		}
+		this.books.remove(0);
+	}
+
+	public List<Book> getLibrary() {
+		return books;
+	}
+	
+	public void printBooks() {
+		for (Book book : this.getLibrary()) {
```
Comment (SerhiiPasko) [inline src/main/java/homework08/BookCrossingLibrary.java]: I many time spoke about this

File: src/main/java/homework08/BringBookThread.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package homework08;
+
+public class BringBookThread extends Thread {
+	protected BookCrossingLibrary bcl;
```
Comment (SerhiiPasko) [inline src/main/java/homework08/BringBookThread.java]: Why protected?

File: src/main/java/homework08/CallElevatorThreadA.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package homework08;
+
+public class CallElevatorThreadA extends Thread{
+
+	protected Elevator elevator = null;
```
Comment (SerhiiPasko) [inline src/main/java/homework08/CallElevatorThreadA.java:5]: Also

File: src/main/java/homework08/Elevator.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+			}
+		}
+		System.out.println("Stop, out");
+		this.floor = goFloor;
+	}
+
+	// Synchronized block
+	public int moveTime(int callFloor, int goFloor) {
+		int floors = 0;
+		synchronized (this) {
```
Comment (SerhiiPasko) [inline src/main/java/homework08/Elevator.java]: Doing this on object not in class level

**Review by SerhiiPasko — CHANGES_REQUESTED** [review]
Read comments. I don't see BlockingQueue

File: src/main/java/homework08/BookCrossingLibrary.java (hunk 1 above)
Comment (un-install) [reply src/main/java/homework08/BookCrossingLibrary.java]: why not to use forEach

File: src/main/java/homework08/BookCrossingLibrary.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package homework08;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class BookCrossingLibrary {
+	private List<Book> books = new LinkedList<Book>();
```
Comment (un-install) [inline src/main/java/homework08/BookCrossingLibrary.java]: why you use LinkedList but not the ArrayList?

File: src/main/java/homework08/BookCrossingLibrary.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	public BookCrossingLibrary(int bookLimit) {
+		this.bookLimit = bookLimit;
+	}
+
+	public synchronized void bringBook(Book item) throws InterruptedException {
+		while (this.books.size() == this.bookLimit) {
+			wait();
+		}
+		if (this.books.size() == 0) {
+			notifyAll();
```
Comment (un-install) [inline src/main/java/homework08/BookCrossingLibrary.java]: this is not thread safe. Imagine that you have some independent waiting threads.

File: src/main/java/homework08/MainForBooks.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+		List<Book> books = Arrays.asList(book1, book2, book3, book4, book5, book6, book7, book8, book9, book10);
+		BookCrossingLibrary lib = new BookCrossingLibrary(10);
+		for (Book book : books) {
+			lib.bringBook(book);
+		}
+		TakeBookThread tbt = new TakeBookThread(lib);
+		BringBookThread bbt = new BringBookThread(lib);
+		tbt.run();
+		bbt.run();
```
Comment (un-install) [inline src/main/java/homework08/MainForBooks.java]: there is no realy need in sync methods if you use only one thread for each

File: src/main/java/homework08/BookCrossingLibrary.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		}
+		this.books.add(item);
+	}
+
+	public synchronized void takeBook() throws InterruptedException {
+		while (this.books.size() == 0) {
+			wait();
+		}
+		if (this.books.size() == this.bookLimit) {
+			notifyAll();
```
Comment (un-install) [inline src/main/java/homework08/BookCrossingLibrary.java]: Also

File: src/main/java/homework08/BringBookThread.java (hunk 2 above)
Comment (un-install) [reply src/main/java/homework08/BringBookThread.java]: breeding of entitys

Comment (u2107680007911543774e7r) [reply src/main/java/homework08/BringBookThread.java]: > breeding of entitys

what does it mean?

Comment (u2107680007911543774e7r) [reply src/main/java/homework08/BringBookThread.java]: > Why protected?

I see protected in example - I do protected in my homework : )
Maybe I will use this in other hw.

**Comment by saadow** [comment]
> Read comments. I don't see BlockingQueue

Bookcrossing is Blocking queue... People bring books so others can take them.

File: src/main/java/homework08/BookCrossingLibrary.java (hunk 5 above)
Comment (u2107680007911543774e7r) [reply src/main/java/homework08/BookCrossingLibrary.java]: > why you use LinkedList but not the ArrayList?

I only need add and remove so LinkedList is better

File: src/main/java/homework08/BookCrossingLibrary.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	private BlockingQueue<Book> books;
+	private int bookLimit = 10;
+
+	public BookCrossingLibrary(int bookLimit) {
+		this.bookLimit = bookLimit;
+	}
+
+	public synchronized void bringBook(Book item) throws InterruptedException {
+		while (this.books.size() == this.bookLimit) {
+			wait();
```
Comment (SerhiiPasko) [inline src/main/java/homework08/BookCrossingLibrary.java]: I dont understand why you need wait and notify

Comment (SerhiiPasko) [reply src/main/java/homework08/BookCrossingLibrary.java]: You have now two synhronize methods. On level BookCrossingLibrary  you can use it same time only one

**Review by SerhiiPasko — CHANGES_REQUESTED** [review]
You use wait and notify in places that doesn't need

**Comment by u2107680007911543774e7r** [comment]
> You use wait and notify in places that doesn't need

I think I read wrong article about Blocking Queue : )

**Pull request merged**